File: Dichotomie in Caculus Test/dichotomie.c
#include <stdio.h>

double f(double x){
    return x*x*x*x*x + 4*x*x*x - 2;
}
int main(){
    double amplitude = (double)1/0xFFFFFF;
    double intervalle[2] = {0.0,1.0};
    printf("amplitude: %g\n",amplitude);
    dichotomie(intervalle,amplitude);
    return 0;
}

void dichotomie(double inv[2], double amp){
    double centre = (inv[0] + inv[1])/2;
    double new_inv[2];
    if((inv[1] - inv[0]) > amp){

        if(f(centre)*f(inv[0]) <0){
            new_inv[0] = inv[0];
            new_inv[1] = centre;
        }
        if(f(centre)*f(inv[1]) <0){
            new_inv[0] = centre;
            new_inv[1] = inv[1];
        }
        //printf("centre: %f\nf(centre): %f\nf(inv[0]): %f\nf(inv[1]): %f\nnew_inv[0]: %f\nnew_inv[1]: %f\n",centre,f(centre),f(inv[0]),f(inv[1]), new_inv[0], new_inv[1]);
        printf("inv 1: %g , inv 2: %g\n",new_inv[0],new_inv[1]);
        dichotomie(new_inv,amp);
    }
}

File: Dichotomie in Caculus Test/Dichotomie.java
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Stream;

class Dichotomie {

    static StringBuilder sb = new StringBuilder();
    static PrintWriter pw;
    static Function<Double, Double> fn;
    static double[] intervalle1 = { 1F/Math.E, 1F/2F };
    double[] new_intervalle = new double[2];

    public static void main(String[] args) throws IOException {
        pw = new PrintWriter("dichotomie.csv");
        fn = x -> x- Math.pow(Math.log(x), 2);
   
        Arrays.stream(new Dichotomie().dichotomie(intervalle1, 1F / 0xFFFFFF, fn)).forEach(System.out::println);
        pw.write("centre,image de centre,intervalle1,intervalle2,amplitude\n");
        pw.write(sb.toString());
        pw.close();
    }

    private double[] dichotomie(double[] intervalle, double amp, Function<Double, Double> function) {
        double center = (intervalle[0] + intervalle[1]) / 2;
        double i0_image = function.apply(intervalle[0]);
        double i1_image = function.apply(intervalle[1]);
        double center_image = function.apply(center);

        if ((intervalle[1] - intervalle[0]) > amp) {
            if (i0_image * center_image < 0) {

                new_intervalle[0] = intervalle[0];
                new_intervalle[1] = center;

            } else if (i1_image * center_image < 0) {

                new_intervalle[0] = center;
                new_intervalle[1] = intervalle[1];

            }
            sb.append(
                    center + "," + center_image + "," + new_intervalle[0] + "," + new_intervalle[1] + "," + (intervalle[1] - intervalle[0]) + "\n");
            dichotomie(new_intervalle, amp, function);
        }
        return new_intervalle;
    }
}